Sources/Features/Client/BacktraceOomWatcher.swift

```diff
@@ -0,0 +1,179 @@
+            _ = try backtraceApi.send(report)
+        } catch {
+            BacktraceLogger.error(error)
+            try? self.repository.save(report)
+        }
+        try! FileManager.default.removeItem(at: BacktraceOomWatcher.oomFilePath!)
+    }
+}
+
+extension BacktraceOomWatcher {
+    /// Describes the current application's state
+    enum ApplicationState: String, Codable {
+        /// The app is in the foreground and actively in use.
+        case active
+        /// The app is in an inactive state when it is in the foreground but receiving events.
+        case inactive
+        /// The app transitions into the background.
+        case background
+    }
+    
+    //// Describes the current application's information
+    struct ApplicationInfo: Codable {
+        var state: ApplicationState = .active
+        var debugger: Bool = DebuggerChecker.isAttached()
+        var appVersion: String = Bundle.main.object(forInfoDictionaryKey: kCFBundleVersionKey as String) as! String
+        var version: String = ProcessInfo.processInfo.operatingSystemVersionString
+        var attributes: Data?
+        var resource: Data?
+    }
+}
+
+extension BacktraceOomWatcher {
+    func appChanedState(_ newState: ApplicationState) {
+        self.state.state = newState
+        saveState()
+    }
+    
+    func handleTermination() {
+        // application terminates correctly - for example: user decide to close app
+        try! FileManager.default.removeItem(atPath: BacktraceOomWatcher.oomFilePath!.path)
+        NotificationCenter.default.removeObserver(self)
+    }
+    
+    func handleLowMemoryWarning() {
+        guard let resource = try? crashReporter.generateLiveReport(exception: nil,
+                                                                   attributes: attributesProvider.allAttributes,
+                                                                   attachmentPaths: []) else {
+                                                                    return
+        }
+        self.state.resource = resource.reportData
+        resource.attributes["error.message"] = "Out of memory detected."
+        resource.attributes["error.type"] = "Low Memory"
+        resource.attributes["state"] = state.state.rawValue
+        
+        
+        self.state.attributes = try? JSONSerialization.data(withJSONObject: resource.attributes)
+        saveState()
+    }
+    
+    private func loadPreviousState() -> ApplicationInfo? {
+        let decoder = PropertyListDecoder()
+        
+        guard let data = try? Data.init(contentsOf: BacktraceOomWatcher.oomFilePath!),
+            let previousAppState = try? decoder.decode(ApplicationInfo.self, from: data)
+            else { return nil }
+        return previousAppState
+        
+    }
+    
+    private func saveState() {
+        let encoder = PropertyListEncoder()
+        if let data = try? encoder.encode(self.state) {
+            let destPath = BacktraceOomWatcher.oomFilePath!.path
```

Comment:
Please avoid force unwrapping

Reply:
avoided unwrapped - added one more if statement. However the one issue that I have with this implementation is that - we know this variable always exist - based on database implementation. We are always sure that oomFilePath will point to right place after runnning the class constructor.